taichi/backends/cuda/cuda_profiler.cpp

```diff
@@ -0,0 +1,150 @@
+#include "taichi/backends/cuda/cuda_profiler.h"
+#include "taichi/backends/cuda/cuda_driver.h"
+#include "taichi/backends/cuda/cuda_context.h"
+
+TLANG_NAMESPACE_BEGIN
+
+CUDAProfiler::CUDAProfiler() {
+  TI_TRACE("CUDAProfiler::CUDAProfiler() ");
+}
+
+CUDAProfiler::~CUDAProfiler() {
+  if (profiler_config_.profiler_type == CUDA_KERNEL_PROFILER_CUPTI) {
+    end_profiling();
+    deinit_cupti();
+  }
+}
+
+bool CUDAProfiler::set_profiling_mode(KernelProfilerMode profiling_mode) {
+  bool enable = profiling_mode == KernelProfilerMode::enable |
+                profiling_mode == KernelProfilerMode::cuda_accurate |
+                profiling_mode == KernelProfilerMode::cuda_detailed;
+  if (!enable) {
+    return false;
+  }
+
+  CUDAKernalProfiler profiler_type =
+      (profiling_mode == KernelProfilerMode::cuda_accurate) ||
+              (profiling_mode == KernelProfilerMode::cuda_detailed)
+          ? CUDA_KERNEL_PROFILER_CUPTI
+          : CUDA_KERNEL_PROFILER_EVENT;
+  profiler_config_.profiling_mode = profiling_mode;
+
+  if (enable && profiler_type == CUDA_KERNEL_PROFILER_EVENT) {
+    profiler_config_.enable = true;
+    profiler_config_.profiler_type = CUDA_KERNEL_PROFILER_EVENT;
+    profiler_config_.profiling_mode = KernelProfilerMode::enable;
+    TI_TRACE("CUDA_KERNEL_PROFILER_EVENT : enable");
+    TI_TRACE("CUDAProfiler enable = {}", profiler_config_.enable);
+    TI_TRACE("profiler_type : {}", profiler_config_.profiler_type);
+    return true;
+  }
+
+#if !defined(TI_WITH_TOOLKIT_CUDA)
```

Comment:
Hmmm it's not clear yet when we should define TI_WITH_TOOLKIT_CUDA, IIUC this should happen in cmake in your next PR when adding CUPTI?